KEYCLOAK-2045 Simple Validation API

thomasdarimont opened the pull request:
Another try for a simple Validator SPI for Keycloak.

This builds on the ideas from #7324 in a lean fashion. This introduces a new Validator SPI that supports built-in as well as user provided validations.

<!---
Please read https://github.com/keycloak/keycloak/blob/master/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

thomasdarimont committed:
KEYCLOAK-2045 Simple Validation API

hmlnarik commented:
Thank you @thomasdarimont. Validation seems to be a rather broad problem in Keycloak spanning several layers, and as such would need a proper design in keycloak-community repo first.
A random thought - can we leverage JSR 303 (Java API for JavaBean Validation)?

thomasdarimont committed:
KEYCLOAK-2045 Return ValidationContext with the outcome of the validation in validate methods

thomasdarimont committed:
KEYCLOAK-2045 Move validator lookup from KeycloakSession into dedicated ValidatorLookup class for now

thomasdarimont commented:
@hmlnarik thanks for the feedback. I've been working on a new validation SPI for quite a while now as you can see here: https://github.com/keycloak/keycloak/pull/7324

Perhaps you should talk to @pedroigor and / or @velias who are also involved in the design of this to ease the work on the user-profile support.

thomasdarimont committed:
KEYCLOAK-2045 Validator SPI polishing

thomasdarimont committed:
KEYCLOAK-2045 Move Validator SPI to server-spi-private for now

thomasdarimont commented:
@pedroigor as discussed I moved the implementation to the server-spi-private module.

The package name `org.keycloak.validate` is a placeholder for now. `org.keycloak.validation` should be the final place, but this requires some adaptation of the existing classes in `org.keycloak.validation` to the new validation api.

thomasdarimont committed:
KEYCLOAK-2045 Add test to demonstrate validator config validation.

thomasdarimont committed:
KEYCLOAK-2045 Add support for context-less validations.

hmlnarik commented:
> @hmlnarik thanks for the feedback. I've been working on a new validation SPI for quite a while now as you can see here: #7324

I can see that as well as that this has been closed without merging.

> Perhaps you should talk to @pedroigor and / or @velias who are also involved in the design of this to ease the work on the user-profile support.

The point is that currently the validation in Keycloak is broken design-wise: there are too many validations scattered across the codebase that are mutually incompatible. So far there has been no consensus as of how to do the validation consistently. This PR seems to add yet another way for the validations. Before this can be considered for merging, the full picture how the existing validations would be transformed into a single approach needs to be clear, and sadly I have not seen any proposal in this regard.

Perhaps you can point me to a design document about that?

Validation is known priority in Keycloak and needs to be done, and needs to be done right. I am not arguing that this PR is not the right approach. However I need to see how the proposed approach would fit existing codebase and cover existing usecases before assessing this PR.

Cc: @stianst @mposolda @vmuzikar @mhajas

thomasdarimont commented:
@hmlnarik Thank you for the detailed feedback.

I am aware that validation is a very complex topic in Keycloak and the area is currently a bit messy.

The current validation SPI proposal is based on numerous discussions on user profiles and other topics on the mailing list and direct meetings.

I am sorry if there were any misunderstandings.

I have just added a PR with a corresponding design document in the Keycloak community repository. 
This document also contains the rationale behind not using Bean Validation.

Hope that helps to clarify a few things.

thomasdarimont committed:
KEYCLOAK-2045 Introduce ValidatorConfig and add some default Validator implementations

thomasdarimont committed:
KEYCLOAK-2045 Add license headers

thomasdarimont committed:
KEYCLOAK-2045 Revise BuiltinValidators

thomasdarimont committed:
KEYCLOAK-2045 Validator Polishing

thomasdarimont committed:
KEYCLOAK-2045 Allow lookups for ValidatorFactories for dynamic config validation.

thomasdarimont committed:
KEYCLOAK-2045 Revise ValidatorTests

thomasdarimont committed:
KEYCLOAK-2045 Validator Polishing

vmuzikar commented:
@hmlnarik +1

I think we should discuss and merge the design doc first.

thomasdarimont committed:
KEYCLOAK-2045 Add example for nested validation

pedroigor commented:
@hmlnarik @vmuzikar Please, also consider looking at https://github.com/keycloak/keycloak/pull/7846.

That is what is driving the design for this validation and how it would be used in contexts other than user profile.

stianst commented:
I'd like to make it fully clear here that we are reviewing this work in the context of user profile only. Before we use this validation framework for other purposes we have some areas like what @hmlnarik has mentioned that requires a bit more thought and consideration before we consider using this as the validation framework for other purposes as well. However, this is a good start and I would say that most likely this will evolve to becoming such a framework in the future, but follow-up both in terms of design and code would be needed first.

thomasdarimont committed:
KEYCLOAK-2045 Refactor Validator API

hmlnarik commented:
Thank you @stianst.
If the proposed solution is expected to only cover user profile and is expected as a short-term solution, then it works for me.
For the reference, the more in-depth solution would likely target validation on the level of `*Model` interfaces (e.g. `UserModel`, `ClientModel`) rather than representations and REST endpoints. That way maintenance cost and space for omissions and inconsistencies between validations of the same type of object in several places would be kept minimal.

thomasdarimont commented:
@hmlnarik Thanks again for the feedback.

I think the proposed Validation API will work fine for validating `*Model` objects.

pedroigor commented:
@thomasdarimont I agree. That is an implementation detail that would basically rely on looking annotations and mapping to their respective validators.

Another point I would highlight is that by having a very minimal and really simple API, we don't need any other dependency added and something more focused on our needs. Although Keycloak is basically a regular JAX-RS application, our requirements usually extrapolate what specs and 3rd party libraries offer in terms of API.

thomasdarimont committed:
KEYCLOAK-2045 Refactor Validator API

- Merged BuiltinValidators and ValidatorLookup into Validators facade class

thomasdarimont committed:
KEYCLOAK-2045 Ease validation of ValidatorConfigs

jbman commented on server-spi-private/src/main/java/org/keycloak/validate/ValidationResult.java:
```diff
@@ -31,6 +31,11 @@
+    /**
+     * An empty ValidationResult that's invalid by default.
+     */
+    public static final ValidationResult INVALID = new ValidationResult(false, Collections.emptySet());
```
Why should it be allowed to have an invalid result but no validation errors? Would be less surprising that an invalid result is always based on validation errors.

jbman commented on server-spi-private/src/main/java/org/keycloak/validate/Validator.java:
```diff
@@ -0,0 +1,114 @@
+     * @param config parameterization for the current validation
+     * @return the validation context with the outcome of the validation
+     */
+    default ValidationContext validate(Object input, ValidatorConfig config) {
```
1. Was it considered to have ValidatorConfig as part of the ValidationContext?
2. The ValidationContext including session should always be available to the validator, e.g for getting other validators and delegate. 
But method `ValidationContext validate(Object input)` allows to request validation without session available. Waht is a validotor supposed to do which needs the session?

thomasdarimont replied:
1. Yes it was, I had this (via simple `Map<String,Object> attributes`) in my initial API design: https://github.com/keycloak/keycloak/pull/7324/files#diff-a30ff7afb40dfeee146a59e5e46fbd15598d5d36a2dbd616095d049e5181613dR36
We want to support simple and complex validations. Simple validations (just one call to validate(...)) don't require an initial `ValidationConfig` but might want to fine tune the `Validator`.
Complex validations might invoke `validate(...)` consecutively for different `Validator`s with a shared `ValidationContext`. Since the `ValidationContext` is shared across multiple `Validator` invocations, it would require additional context management to maintain the proper `ValidationConfig`. Explicitly passing the `ValidationConfig` makes the API easier for simple and complex validations.

2. A `Validator` which needs the `KeycloakSession` needs to be invoked with a properly configured `ValidationContext` with a session. `#validate(object)` allows simple validations like
```
Assert.assertTrue(Validators.email().validate("admin@example.org").isValid());
```
while still sharing the common `Validator` interface.

jbman replied:
1. Thanks, shared context but not shared config is a good reason to separate the parameters
2. Ok, got the different cases. The main drawback is the difficulty for the validator implementation to state that that  a 'ValidationContext' instance is required and must not be null.  Would be nice to know that just by the interface. 
```
public interface Validator {
    ValidationContext validate(Object input, ValidatorConfig config)) ...

public interface ValidatorWithContext {
    ValidationContext validate(Object input, ValidationContext context, ValidatorConfig config)) ...
```
Unfortunately this can't be done for 'ValidatorConfig' parameter at the same time. Get's a bit unhandy to have a 'ValidatorWithContextAndConfig' and a 'ValidatorWithConfig' interface on top.

thomasdarimont committed:
KEYCLOAK-2045 Assume ValidatorConfig's are valid for Validator's that don't support config.

thomasdarimont commented on server-spi-private/src/main/java/org/keycloak/validate/builtin/LengthValidator.java:
```diff
@@ -0,0 +1,120 @@
+        }
+
+        String string = (String) input;
+        int min = config.getIntOrDefault(KEY_MIN, 0);
```
That was indeed my initial design, but then the Keycloak team asked how validator configuration can be validated with this approach. That's why I introduced the `ValidatorFactory.validateConfig(..)` method.

The `ValidatorFactory.validateConfig(..)` method allows to validate `Validator` configurations during edit in the admin-console.

thomasdarimont committed:
KEYCLOAK-2045 Make built-in validator method names more for static imports

vmuzikar commented on server-spi-private/src/main/java/org/keycloak/validate/builtin/EmailValidator.java:
```diff
@@ -0,0 +1,60 @@
+    public ValidationContext validate(Object input, String inputHint, ValidationContext context, ValidatorConfig config) {
+
+        if (!(input instanceof String)) {
+            context.addError(new ValidationError(ID, inputHint, MESSAGE_INVALID_EMAIL, input));
```
Should we rather throw some exception?  It's not exactly a validation error, the user providing the input is not responsible for this.

thomasdarimont replied:
I'd rather not throw an exception by default in that case, but we could configure the ValidationContext to throw an exception when a ValidationError is reported.

vmuzikar replied:
Ok. My original assumption was that the user cannot affect the input type and therefore throwing incompatible input type at the validator is a dev's fault and as such is not a validation error. This is, of course, wrong – a user can in fact define input type (e.g. JSON payload in REST endpoints). Let's return a validation error.

But maybe we could have some predefined error message common for all validators for this case (incompatible input type)?

vmuzikar commented on server-spi-private/src/main/java/org/keycloak/validate/builtin/LengthValidator.java:
```diff
@@ -0,0 +1,120 @@
+    @Override
+    public ValidationContext validate(Object input, String inputHint, ValidationContext context, ValidatorConfig config) {
+
+        if (input == null) {
+            context.addError(new ValidationError(ID, inputHint, MESSAGE_INVALID_LENGTH, input));
+            return context;
+        }
+
+        if (!(input instanceof String)) {
+            return context;
+        }
```
Should we rather throw some exception?

thomasdarimont replied:
As said above, we could report a  ValidationError here.

vmuzikar commented on server-spi-private/src/main/java/org/keycloak/validate/builtin/LengthValidator.java:
```diff
@@ -0,0 +1,120 @@
+        if (containsMax && !(maybeMax instanceof Integer)) {
+            errors.add(new ValidationError(ID, KEY_MAX, MESSAGE_INVALID_VALUE, maybeMax));
+        }
+
```
Should we check `min <= max`?

thomasdarimont replied:
Yes

vmuzikar commented on server-spi-private/src/main/java/org/keycloak/validate/builtin/NotBlankValidator.java:
```diff
@@ -0,0 +1,55 @@
+    @Override
+    public ValidationContext validate(Object input, String inputHint, ValidationContext context, ValidatorConfig config) {
+
+        if (!(input instanceof String)) {
+            context.addError(new ValidationError(ID, inputHint, MESSAGE_INVALID_VALUE, input));
+            return context;
+        }
```
Should we rather throw some exception?

vmuzikar commented on server-spi-private/src/main/java/org/keycloak/validate/ValidatorConfig.java:
```diff
@@ -0,0 +1,184 @@
+        if (value instanceof Pattern) {
+            return (Pattern) value;
+        } else if (value instanceof String) {
+            return Pattern.compile((String) value);
```
Should we store the compiled pattern for the next time?

thomasdarimont replied:
I'd delegate that to the ValidatorConfig. But yes, the pattern could be cached.

vmuzikar commented on server-spi-private/src/main/java/org/keycloak/validate/builtin/PatternValidator.java:
```diff
@@ -0,0 +1,82 @@
+    @Override
+    public ValidationContext validate(Object input, String inputHint, ValidationContext context, ValidatorConfig config) {
+
+        if (!(input instanceof String)) {
+            context.addError(new ValidationError(ID, inputHint, MESSAGE_INVALID_VALUE, input));
+            return context;
+        }
```
Should we rather throw some exception?

vmuzikar commented on server-spi-private/src/main/java/org/keycloak/validate/builtin/LengthValidator.java:
```diff
@@ -0,0 +1,120 @@
+    @Override
+    public ValidationContext validate(Object input, String inputHint, ValidationContext context, ValidatorConfig config) {
+
+        if (input == null) {
+            context.addError(new ValidationError(ID, inputHint, MESSAGE_INVALID_LENGTH, input));
+            return context;
+        }
```
I noticed in general some inconsistencies in the built-in validators in how the validation is performed if input is null/empty. I think we might want to skip validations by default if given input is null/empty. Non-empty validation should be probably done explicitly if needed. Or maybe this could be somehow configurable in the context?

thomasdarimont replied:
I agree, the current implementations are WIP and needs to be aligned. I like your idea of handling null values.

velias replied:
For user profile, we need all these "format" validations to run only if value is not null nor empty

vmuzikar reviewed (changes requested):
@thomasdarimont Nice work, thanks!

thomasdarimont commented on server-spi-private/src/main/java/org/keycloak/validate/Validator.java:
```diff
@@ -0,0 +1,114 @@
+ * <p>
+ * Validations can be configured with an optional {@code config} {@link Map}.
+ */
+public interface Validator extends Provider {
```
A SAM type interface allow validator definitions from a lambda expression. That was an initial recommendation of Stian

thomasdarimont commented on server-spi-private/src/main/java/org/keycloak/validate/ValidationContext.java:
```diff
@@ -0,0 +1,121 @@
+        return Validators.validator(session, validatorId);
+    }
+
+    public void addError(ValidationError error) {
+        errors.add(error);
+    }
```
I considered that option initially, but postponed this as I wanted to keep the API simple at first to ease discussions :)

pedroigor replied:
+1 we can always improve that (and other places) once we start using and identifying boilerplate code.

thomasdarimont commented on server-spi-private/src/main/java/org/keycloak/validate/builtin/EmailValidator.java:
```diff
@@ -0,0 +1,60 @@
+
+    // Actually allow same emails like angular. See ValidationTest.testEmailValidation()
+    private static final Pattern EMAIL_PATTERN = Pattern.compile(
+            "[a-zA-Z0-9!#$%&'*+/=?^_`{|}~.-]+@[a-zA-Z0-9-]+(\\.[a-zA-Z0-9-]+)*");
```
I used th regex that is used currently in Keycloal :), the expression can be adapted of course.

vmuzikar replied:
Ah, right. :D Let's leave it as is then.

thomasdarimont commented on server-spi-private/src/main/java/org/keycloak/validate/builtin/NotEmptyValidator.java:
```diff
@@ -0,0 +1,74 @@
+            return context;
+        }
+
+        return context;
```
For situations where a validator does not support the input, we could emit an ValidationError with the MESSAGE_INVALID_VALUE message. This message can also be used if an Validator config option is used that is not supported.

thomasdarimont committed:
KEYCLOAK-2045 Revise Validation API according to discussions from code review

thomasdarimont committed:
KEYCLOAK-2045 Pass KeycloakSession to ValidatorConfig validation methods.

This allows ValidatorConfig validations to access services as well as
the current KeycloakContext.

pedroigor commented on server-spi-private/src/main/java/org/keycloak/validate/builtin/NumberValidator.java:
```diff
@@ -0,0 +1,71 @@
+            // try to parse the string into a number
+            String string = (String) input;
+            try {
+                Integer.parseInt(string);
```
Perhaps just check if can be parsed as a double?

thomasdarimont replied:
Okay

pedroigor reviewed (commented):
@thomasdarimont Looks great. Just a few minor comments.

thomasdarimont commented on server-spi-private/src/main/java/org/keycloak/validate/ValidationResult.java:
```diff
@@ -0,0 +1,149 @@
+    /**
+     * Holds the validation status.
+     */
+    private final boolean valid;
```
I agree, I will change that.

thomasdarimont commented on server-spi-private/src/main/java/org/keycloak/validate/ValidationResult.java:
```diff
@@ -0,0 +1,149 @@
+/**
+ * Denotes the result of a validation.
+ */
+public class ValidationResult implements Consumer<Consumer<ValidationResult>> {
```
The rational was to make it easier to use `ValidationResult` in fluent API call chains. I just updated `ValidationResult` to no longer inherit from `Consumer<Consumer<ValidationResult>>`.

thomasdarimont commented on server-spi-private/src/main/java/org/keycloak/validate/ValidatorConfig.java:
```diff
@@ -0,0 +1,184 @@
+        if (value instanceof Pattern) {
+            return (Pattern) value;
+        } else if (value instanceof String) {
+            return Pattern.compile((String) value);
```
I agree, that a pre-compiled pattern is preferred here. The rational for the string based pattern here is the support for profile validation config from a JSON string.

But, of course we could also compile the pattern during the conversion of the json based profile validation config into a `ValidatorConfig`.

I'll remove the support for a string based pattern.

thomasdarimont commented:
Just rebased this on current master.

thomasdarimont committed:
KEYCLOAK-2045 Revise Validation API according to discussions from code review

- Derive valid state from errors in ValidationContext
- Simplify ValidationResult by not inheriting from Consumer<Consmer<ValidationResult>>
- Add addtional test cases

thomasdarimont committed:
KEYCLOAK-2045 Revise Validation API according to discussions from code review

- Revise NumberValidator to only check for double conversion
- Removed support for string based regex pattern

thomasdarimont committed:
KEYCLOAK-2045 Ease validation of ValidatorConfig

pedroigor reviewed (dismissed):
@thomasdarimont Thanks.

@hmlnarik @vmuzikar LGTM.

thomasdarimont commented:
Thanks for having another look at this :)

I was wondering what the next steps will be.

In my point of view the next steps should address the following points: 
- merge the `validate` package into the `validation` package to avoid confusing package names
- refactor the existing validation package to use the new validation api.
- adapt the profile api to use the new validation api (create validation configs from json based profile definitions)
- add / revise built-in validations as needed

pedroigor commented:
> Thanks for having another look at this :)
> 
> I was wondering what the next steps will be.
> 
> In my point of view the next steps should address the following points:
> 
> * merge the `validate` package into the `validation` package to avoid confusing package names
> * refactor the existing validation package to use the new validation api.
> * adapt the profile api to use the new validation api (create validation configs from json based profile definitions)
> * add / revise built-in validations as needed

Yeah, pretty much what we discussed. Your proposal should be able to replace the different validation approaches we have but I'm not sure if we want to do that now but focus on the user profile bits first.

We are already changing a lot in the user profile and the product release is quite close. So perhaps we should consider changing other parts when we have more time to bake their changes. In fact, I have the same concern surrounding the user profile, but this one is something we need ASAP. I'm also more confident that we have much better coverage for making deep changes in user profile after the last issues we had.

That said, I would focus on user profile for now.

velias commented on server-spi-private/src/main/java/org/keycloak/validate/builtin/LengthValidator.java:
```diff
@@ -0,0 +1,122 @@
+        if (!(containsMin || containsMax)) {
+            return ValidationResult.OK;
+        }
+
```
I'd say that this validator needs at least one of `min` or `max` to be set in the configuration to work properly. So this config validation should fail if config is empty or doesn't contain any of these settings.

velias commented on server-spi-private/src/main/java/org/keycloak/validate/builtin/EmailValidator.java:
```diff
@@ -0,0 +1,60 @@
+    @Override
+    public ValidationContext validate(Object input, String inputHint, ValidationContext context, ValidatorConfig config) {
+
+        if (!(input instanceof String)) {
```
User Profile Attributes are List of Strings. We should probably implement all format validators to support lists also, validate every item in the list and report error even if one of them doesn't match.

velias commented on server-spi-private/src/main/java/org/keycloak/validate/builtin/EmailValidator.java:
```diff
@@ -0,0 +1,60 @@
+            return context;
+        }
+
+        if (!EMAIL_PATTERN.matcher((String) input).matches()) {
```
This is format validation, so value should be validated only if it is not null to be able to operate on optional fields.

pedroigor commented on services/src/main/java/org/keycloak/services/DefaultKeycloakSession.java:
```diff
@@ -389,8 +388,8 @@ public UserCredentialManager userCredentialManager() {
     @Override
     public <T extends Provider> Set<T> getAllProviders(Class<T> clazz) {
```
Can we revert changes to this class?

pedroigor commented on server-spi/src/main/java/org/keycloak/models/KeycloakSession.java:
```diff
@@ -20,7 +20,6 @@
 import org.keycloak.component.ComponentModel;
 import org.keycloak.models.cache.UserCache;
 import org.keycloak.provider.InvalidationHandler;
-import org.keycloak.provider.InvalidationHandler.InvalidableObjectType;
```
Can we revert changes to this class?

pedroigor commented:
@thomasdarimont I think we should change the built-in validators to handle multiple values too. Would you be able to update them?

pedroigor commented:
@thomasdarimont As discussed, we are closing this one and sending a new one with the few missing bits. Thanks for your work.
